File: src/main/java/dev/slimevr/vr/poserecorder/TrackerFrame.java
package dev.slimevr.vr.poserecorder;

import com.jme3.math.Quaternion;
import com.jme3.math.Vector3f;

import io.eiren.vr.processor.TrackerBodyPosition;
import io.eiren.vr.trackers.Tracker;
import io.eiren.vr.trackers.TrackerConfig;
import io.eiren.vr.trackers.TrackerStatus;

public final class TrackerFrame implements Tracker {

	private int dataFlags = 0;

	public final String name;
	public final TrackerBodyPosition designation;
	public final Quaternion rotation;
	public final Vector3f position;

	public TrackerFrame(String designationString, TrackerBodyPosition designation, Quaternion rotation, Vector3f position) {
		if (designationString != null) {
			this.name = "TrackerFrame:/" + designationString;
		} else {
			this.name = "TrackerFrame";
		}

		this.designation = designation;
		if (designation != null) {
			dataFlags |= TrackerFrameData.DESIGNATION.flag;
		}

		this.rotation = rotation;
		if (rotation != null) {
			dataFlags |= TrackerFrameData.ROTATION.flag;
		}

		this.position = position;
		if (position != null) {
			dataFlags |= TrackerFrameData.POSITION.flag;
		}
	}

	public TrackerFrame(TrackerBodyPosition designation, Quaternion rotation, Vector3f position) {
		this(designation == null ? null : designation.designation, designation, rotation, position);
	}

	public TrackerFrame(TrackerBodyPosition designation, Quaternion rotation) {
		this(designation, rotation, null);
	}

	public TrackerFrame(TrackerBodyPosition designation, Vector3f position) {
		this(designation, null, position);
	}

	public static TrackerFrame fromTracker(Tracker tracker) {
		if (tracker == null) {
			return null;
		}

		// If the tracker is not ready
		if (tracker.getStatus() != TrackerStatus.OK && tracker.getStatus() != TrackerStatus.BUSY && tracker.getStatus() != TrackerStatus.OCCLUDED) {
			return null;
		}

		boolean hasPosition = tracker.hasPosition() && tracker.getStatus() != TrackerStatus.OCCLUDED;
		// If tracker has no data
		if (tracker.getBodyPosition() == null && !tracker.hasRotation() && !hasPosition) {
			return null;
		}

		Quaternion rotation = null;
		if (tracker.hasRotation()) {
			rotation = new Quaternion();
			tracker.getRotation(rotation);
		}

		// If the tracker is occluded, don't get the position
		Vector3f position = null;
		if (hasPosition) {
			position = new Vector3f();
			tracker.getPosition(position);
		}

		return new TrackerFrame(tracker.getBodyPosition(), rotation, position);
	}

	public int getDataFlags() {
		return dataFlags;
	}

	public boolean hasData(TrackerFrameData flag) {
		return flag.check(dataFlags);
	}

	//#region Tracker Interface Implementation
	@Override
	public boolean getRotation(Quaternion store) {
		if (hasData(TrackerFrameData.ROTATION)) {
			store.set(rotation);
			return true;
		}

		store.set(0, 0, 0, 1);
		return false;
	}

	@Override
	public boolean getPosition(Vector3f store) {
		if (hasData(TrackerFrameData.POSITION)) {
			store.set(position);
			return true;
		}

		store.set(0, 0, 0);
		return false;
	}

	@Override
	public String getName() {
		return name;
	}

	@Override
	public TrackerStatus getStatus() {
		return TrackerStatus.OK;
	}

	@Override
	public void loadConfig(TrackerConfig config) {
		throw new UnsupportedOperationException("TrackerFrame does not implement configuration");
	}

	@Override
	public void saveConfig(TrackerConfig config) {
		throw new UnsupportedOperationException("TrackerFrame does not implement configuration");
	}

	@Override
	public float getConfidenceLevel() {
		return 0;
	}

	@Override
	public void resetFull(Quaternion reference) {
		throw new UnsupportedOperationException("TrackerFrame does not implement calibration");
	}

	@Override
	public void resetYaw(Quaternion reference) {
		throw new UnsupportedOperationException("TrackerFrame does not implement calibration");
	}

	@Override
	public void tick() {
		throw new UnsupportedOperationException("TrackerFrame does not implement this method");
	}

	@Override
	public TrackerBodyPosition getBodyPosition() {
		return designation;
	}

	@Override
	public void setBodyPosition(TrackerBodyPosition position) {
		throw new UnsupportedOperationException("TrackerFrame does not allow setting the body position");
	}

	@Override
	public boolean userEditable() {
		return false;
	}

	@Override
	public boolean hasRotation() {
		return hasData(TrackerFrameData.ROTATION);
	}

	@Override
	public boolean hasPosition() {
		return hasData(TrackerFrameData.POSITION);
	}

	@Override
	public boolean isComputed() {
		return true;
	}
	//#endregion
}

File: server/core/src/main/java/dev/slimevr/poseframeformat/trackerdata/TrackerFrame.kt
package dev.slimevr.poseframeformat.trackerdata

import dev.slimevr.tracking.trackers.Tracker
import dev.slimevr.tracking.trackers.TrackerPosition
import dev.slimevr.tracking.trackers.TrackerStatus
import io.github.axisangles.ktmath.Quaternion
import io.github.axisangles.ktmath.Vector3

data class TrackerFrame(
	val trackerPosition: TrackerPosition? = null,
	val rotation: Quaternion? = null,
	val position: Vector3? = null,
	val acceleration: Vector3? = null,
	val rawRotation: Quaternion? = null,
) {
	val dataFlags: Int

	val name: String
		get() = "TrackerFrame:/${trackerPosition?.designation ?: "null"}"

	init {
		var initDataFlags = 0

		if (trackerPosition != null) {
			initDataFlags = TrackerFrameData.TRACKER_POSITION_ENUM.add(initDataFlags)
		}
		if (rotation != null) {
			initDataFlags = TrackerFrameData.ROTATION.add(initDataFlags)
		}
		if (position != null) {
			initDataFlags = TrackerFrameData.POSITION.add(initDataFlags)
		}
		if (acceleration != null) {
			initDataFlags = TrackerFrameData.ACCELERATION.add(initDataFlags)
		}
		if (rawRotation != null) {
			initDataFlags = TrackerFrameData.RAW_ROTATION.add(initDataFlags)
		}

		dataFlags = initDataFlags
	}

	fun hasData(flag: TrackerFrameData): Boolean = flag.check(dataFlags)

	// region Tracker Try Getters
	fun tryGetTrackerPosition(): TrackerPosition? = if (hasData(TrackerFrameData.TRACKER_POSITION_ENUM) || hasData(TrackerFrameData.DESIGNATION_STRING)) {
		trackerPosition
	} else {
		null
	}

	fun tryGetRotation(): Quaternion? = if (hasData(TrackerFrameData.ROTATION)) {
		rotation
	} else {
		null
	}

	fun tryGetRawRotation(): Quaternion? = if (hasData(TrackerFrameData.RAW_ROTATION)) {
		rawRotation
	} else {
		null
	}

	fun tryGetPosition(): Vector3? = if (hasData(TrackerFrameData.POSITION)) {
		position
	} else {
		null
	}

	fun tryGetAcceleration(): Vector3? = if (hasData(TrackerFrameData.ACCELERATION)) {
		acceleration
	} else {
		null
	}

	fun hasRotation(): Boolean = hasData(TrackerFrameData.ROTATION)

	fun hasPosition(): Boolean = hasData(TrackerFrameData.POSITION)

	fun hasAcceleration(): Boolean = hasData(TrackerFrameData.ACCELERATION)
	// endregion

	companion object {
		val empty = TrackerFrame()

		fun fromTracker(tracker: Tracker): TrackerFrame? {
			// If the tracker is not ready
			if (tracker.status != TrackerStatus.OK && tracker.status != TrackerStatus.BUSY && tracker.status != TrackerStatus.OCCLUDED) {
				return null
			}

			val trackerPosition = tracker.trackerPosition

			// If tracker has no data at all, there's no point in writing a frame
			// Note: This includes rawRotation because of `!tracker.hasRotation`
			if (trackerPosition == null && !tracker.hasRotation && !tracker.hasPosition && !tracker.hasAcceleration) {
				return null
			}

			val rotation: Quaternion? = if (tracker.hasRotation) tracker.getRotation() else null
			val position: Vector3? = if (tracker.hasPosition) tracker.position else null
			val acceleration: Vector3? = if (tracker.hasAcceleration) tracker.getAcceleration() else null

			var rawRotation: Quaternion? = if (tracker.hasAdjustedRotation) tracker.getRawRotation() else null
			// If the rawRotation is the same as rotation, there's no point in saving it, set it back to null
			if (rawRotation == rotation) rawRotation = null

			return TrackerFrame(
				trackerPosition,
				rotation,
				position,
				acceleration,
				rawRotation,
			)
		}
	}
}
